src/plugins/lexer/mod.ts

```diff
@@ -0,0 +1,390 @@
+// import some types/const enums to keep source tidy
+import { Mode, Token, Trace } from "./types/mod.ts";
+
+/*
+note:
+  lexer uses zero-based numbering for tracking
+  its position and the position of tokens.
+*/
+export class Lexer {
+  source: string = "";
+  // lexer position on file
+  position: number = -1;
+  // virtual position; used for tracking token positions.
+  virtual_position: number = -1;
+  // keeps track of elementary debug info useful for parser
+  trace: Trace = {
+    line: 1,
+    column: [0, 0],
+  };
+  // stores current built token
+  token: string = "";
+  // stores current lexical tokens
+  tokens: Token[] = [];
+  // stores mode lexer is interpreting in
+  mode: Mode = Mode.Token;
+
+  // optional initiator for .mtl source
+  constructor(input?: string) {
+    if (input !== undefined) {
+      this.source = input.replace(/(\r\n|\n|\r)/gm, "&eol") + "&eof";
+    }
+  }
+
+  // loads new .mtl source, avoiding reinitializing class
+  load(input: string) {
+    this.source = input.replace(/(\r\n|\n|\r)/gm, "&eol") + "&eof";
+    return this;
+  }
```

Comment:
This seems redundant. I think a better approach would be to remove the constructor / optional loading method, and instead require the use of load(). This creates uniform calls and removes ambiguity. 
```suggestion
  // loads new .mtl source, avoiding reinitializing class
  load(input: string) {
    this.source = input.replace(/(\r\n|\n|\r)/gm, "&eol") + "&eof";
    return this;
  }
```

Reply:
100% agree. Will be removed in next commit 👍